File: Chapter 10/Python/section_10_3_resilience_decorators/resilient/core/ports/message_queue.py
from abc import ABC, abstractmethod
from ..domain.order_status import OrderStatus

class MessageQueue(ABC):
    """
    THE CORE PORT (The Asynchronous Interface)
    
    @description
    This abstract base class defines the system's ability to defer work. 
    It is a "Primary Boundary" that allows the Business Logic to offload 
    transactions when synchronous gateways (like FlakyPayments) are unreachable.
    
    PRODUCTION REALITY (How this works on a Server):
    While our local adapter uses SQLite, this Port represents the 
    standard 'Producer' pattern found in distributed systems:
    
    1. INDEPENDENCE: The Business Logic calls 'enqueue' and immediately 
       receives a return. It does not wait for the payment to actually 
       be processed.
       
    2. RELIABILITY: In production (RabbitMQ/SQS), this Port ensures that 
       once a message is accepted by the broker, it is stored on multiple 
       disks/nodes so it cannot be lost if a single server crashes.
       
    3. LOAD LEVELING: This Port acts as a 'Buffer.' If 10,000 orders 
       arrive at once, they sit safely in the Queue Port until the 
       Background Workers can process them at a steady pace.
    """

    @abstractmethod
    def enqueue(self, order_id: str, amount: float, status: OrderStatus, key: str):
        """
        Defines the requirement for persisting an order for later processing.
        
        :param order_id: The unique business identifier for the order.
        :param amount: The financial value of the transaction.
        :param status: The current state (e.g., PENDING_PAYMENT).
        :param key: The Idempotency Key to prevent double-charging during recovery.
        """
        pass

File: Chapter 10/JS/section_10_3_resilience_decorators/resilient/core/ports/messageQueue.js
/**
 * THE CORE PORT (The Reactive Airlock):
 * * @description
 * This port defines the system's capability to "defer work" using RxJS Observables. 
 * It represents an asynchronous boundary where the Core can offload data to 
 * protect the user experience from infrastructure instability.
 * * ARCHITECTURAL CRITIQUE:
 * 1. DECOUPLING: By returning an Observable, the Core remains unaware of 
 * the physical implementation (SQLite, RabbitMQ, or SQS). It only cares 
 * about the stream's completion or error.
 * * 2. BACKPRESSURE: Unlike a Promise, an Observable stream allows the system 
 * to handle backpressure and stream cancellation natively if the 
 * underlying adapter supports it.
 * * 3. PRODUCTION REALITY: In a real server environment, this 'enqueue' 
 * represents a handoff to a persistent broker. Once the stream emits 
 * a value, the Core is guaranteed the data is safely outside the 
 * application process.
 */
export class MessageQueue {
    /**
     * @param {Object} payload - The transaction data to be secured.
     * @returns {import('rxjs').Observable<void>}
     */
    enqueue(payload) {
        throw new Error("Port Method 'enqueue' must be implemented by an Adapter.");
    }
}